### PR TITLE
Add X11 support

### PR DESCRIPTION
**Pull request by kazzix14**
Hi. I'm the guy from [here](https://github.com/rust-windowing/winit/issues/1434#issuecomment-751793973).  
I really want an overlay feature and this repo looks good to me.  
Because It seems there's no support other than Windows currently, I added a support for X11.  
This PR also removes some lines in the example so that It gonna be a bit easy to read.

This is how it works on my machine (debian with xfce4)
overlay

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 /target
 **/*.rs.bk
-Cargo.lock
\ No newline at end of file
+Cargo.lock
+/.vscode
```
Comment (maroider) [inline .gitignore]:
```suggestion
```
Please don't add editor-specific entries to the repository's gitignore file.
You should use something like [`core.excludesFile`](https://git-scm.com/docs/git-config#Documentation/git-config.txt-coreexcludesFile) instead.

File: Cargo.toml [hunk 2]
```diff
@@ -14,12 +14,19 @@ categories = ["gui"]
...
+
 [dependencies]
-winit = "0.23.0"
+winit = "0.24.0"
 
-[target.'cfg(windows)'.dependencies]
+[target.'cfg(target_os = "windows")'.dependencies]
 winapi = { version = "0.3.8", features = ["windef", "winuser"] }
 
+[target.'cfg(target_os = "linux")'.dependencies]
```
Comment (maroider) [inline Cargo.toml]:
```suggestion
[target.'cfg(any(target_os = "linux", target_os = "dragonfly", target_os = "freebsd", target_os = "openbsd", target_os = "netbsd"))'.dependencies]
```
I think it's best to not exclude any of the platforms Winit already supports, even if the only one of these platforms I'm even vaguely familiar with is Linux.

File: Cargo.toml [hunk 3]
```diff
@@ -14,12 +14,19 @@ categories = ["gui"]
...
+xcb = { version = "0.9.0", features = ["composite"], optional = true }
```
Comment (maroider) [inline Cargo.toml:28]: Is there a particular reason you chose XCB over Xlib?
Notably, the Rust XCB bindings don't support dynamically linking XCB, which is what I'd prefer to do **if** implementing this is possible with both Xlib and XCB.

**Review by maroider — CHANGES_REQUESTED** [review]
I'd like the platform-specific code to be organized like how [Winit does it](https://github.com/rust-windowing/winit/blob/master/src/platform_impl/mod.rs). `os.rs` exists because I was lazy when I initially made this.

I'm curious about what you want an overlay for. Is it for a game (or games in general) or is it something else?

File: Cargo.toml (hunk 3 above)
Comment (kazzix14) [reply Cargo.toml:28]: At first, I try to use Xlib. But I don't see any bindings around Shape Extension to use with Xlib in [the crate which winit uses](https://martindev.github.io/martock/doc/x11_dl/index.html) or any other crates.  
On the other hand, I can find the one to use with XCB. That's the reason I choose XCB.
I'm not familiar with X11, so there might be another way to implement this future using Xlib.

Comment (maroider) [reply Cargo.toml:28]: Alright. Supporting both isn't something I'd consider necessary at the moment.
I wasn't really expecting to get X11 support in this crate any time soon, anyway, as my attention has been elsewhere.

**Comment by kazzix14** [comment]
Thank you for reviewing and suggestions.  
I applied your suggestions and make a directory structure to be like winit.


> I'm curious about what you want an overlay for. Is it for a game (or games in general) or is it something else?  

I'm creating an application that flows messages sent from the browser on the screen. I've thought the application is for online presentations, but it can be used for game streaming I guess? anyway, I need an overlay for that.  

Btw, I think it's good to reveal inner functions (make_window_overlay(), make_window_overlay_clickthrough(), make_window_overlay_clickable(), set_window_overlay_opacity()).  
If the functions can be used from the outside, We can make windows that other crates create overlay.  

I patched them to `pub fn` on the local and use them to create overlay with glium.  
glium requires WindowBuilder to create display, so I can't use overlay.window().
```rust
    let display = glium::Display::new(wb, cb, &events_loop).unwrap();

    {
        let gl_window = display.gl_window();
        let window = gl_window.window();
        overlay::make_window_overlay(&window);
        overlay::set_window_overlay_opacity(&window, 255);
    }
```

File: src/platform_impl/mod.rs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+pub use self::platform::*;
+
+#[cfg(target_os = "windows")]
```
Comment (maroider) [inline src/platform_impl/mod.rs:3]:
```suggestion
#[cfg(target_os = "windows")]
#[path = "windows/mod.rs"]
```

**Review by maroider — CHANGES_REQUESTED** [review]
I'd like you to run `cargo fmt` before committing. Current `master` is already formatted this way.

> I'm creating an application that flows messages sent from the browser on the screen. I've thought the application is for online presentations, but it can be used for game streaming I guess? anyway, I need an overlay for that.

Is the overlay you want to make intended to cover the entire monitor or just a single non-fullscreen window? I've been meaning to add support for "window-hugging" overlay windows, where the overlay window follows the shape and position of a particular window.

> Btw, I think it's good to reveal inner functions (make_window_overlay(), make_window_overlay_clickthrough(), make_window_overlay_clickable(), set_window_overlay_opacity()).

I think an extension trait would be better, like `WindowExtOverlay`, which would then allow you to use these methods.

In any case, that should be a separate PR.
This PR should be about X11 support.

EDIT: I'm probably going to have to re-think some other aspects of the current public API as well.

**Comment by kazzix14** [comment]
I'm sorry about the formatting. I'll be careful about it.  

> Is the overlay you want to make intended to cover the entire monitor or just a single non-fullscreen window? I've been meaning to add support for "window-hugging" overlay windows, where the overlay window follows the shape and position of a particular window.

It's intended to cover the entire monitor, but I think It's really great that If it could be a  "window-hugging" overlay window. Because users can use windows that are not wanted to be seen at the same time showing incoming messages and a target window.  
I also play games. If the overlay window can follow my game window, there's a lot of comforts I think.

I totally agree with your opinion about the trait.

**Review by maroider — APPROVED** [review]

**Comment by maroider** [comment]
I did a manual merge.

**Comment by maroider** [comment]
And I apparently suck at git.

~~I reverted the Winit upgrade so that the example continues to compile.~~
Turns out my Cargo.lock was outdated.

**Pull request closed without merging**